pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,380 @@
+	PodsByIPIndexName = "PodByIP"
+	// AnnotationPrefix is the annotation prefix that is turned into tags by default.
+	AnnotationPrefix = "gostatsd.atlassian.com/"
+	// NodeNameEnvVar is the name of the environment variable that the k8s cloud provider will find it's node name in.
+	NodeNameEnvVar = "GOSTATSD_NODE_NAME"
+
+	// ParamAPIQPS is the maximum amount of queries per second we allow to the Kubernetes API server. This is
+	// so we don't overwhelm the server with requests under load.
+	ParamAPIQPS = "kube-api-qps"
+	// ParamAPIQPSBurstFactor is the amount of queries per second we can burst above ParamAPIQPS.
+	ParamAPIQPSBurstFactor = "kube-api-burst"
+	// ParamAnnotationTagWhitelist is a list of regexes to check annotations against. Any pod annotations matching
+	// this pattern will be included as tags on metrics emitted by that pod.
+	ParamAnnotationTagWhitelist = "annotation-tag-whitelist"
+	// ParamLabelTagWhitelist is a list of regexes to check labels against. Any pod labels matching this
+	// pattern will be included as tags on metrics emitted by that pod.
+	ParamLabelTagWhitelist = "label-tag-whitelist"
+	// KubeconfigContextis the name of the context to use inside a provided ParamKubeconfigPath. If ParamKubeconfigPath
+	// is unset this has no effect.
+	ParamKubeconfigContext = "kubeconfig-context"
+	// ParamKubeconfigPath is the path to the kubeconfig file to use for auth, or none if using in-cluster auth.
+	ParamKubeconfigPath = "kubeconfig-path"
+	// ParamResyncPeriod is the resync period for the pod cache as a Duration.
+	ParamResyncPeriod = "resync-period"
+	// ParamWatchCluster is true if we should watch pods in the entire cluster, false if we should watch pods on our
+	// own node.
+	ParamWatchCluster = "watch-cluster"
+	// UserAgent is the base user agent used when talking to the k8s API. A version is appended to this at runtime.
+	UserAgent = "gostatsd"
+
+	// DefaultAPIQPS is the default maximum amount of queries per second we allow to the Kubernetes API server.
+	DefaultAPIQPS = 5
+	// DefaultAPIQPSBurstFactor is the default amount of queries per second we can burst above the maximum QPS.
+	DefaultAPIQPSBurstFactor = 1.5
+	// DefaultKubeconfigContext is the default context to use inside the kubeconfig file. None means use the current
+	// context without switching.
+	DefaultKubeconfigContext = ""
+	// DefaultKubeconfigPath is the default path to the kubeconfig file. None means use in-cluster auth.
+	DefaultKubeconfigPath = ""
+	// DefaultResyncPeriod is the default resync period for the pod cache.
+	DefaultResyncPeriod = 5 * time.Minute
+	// DefaultWatchCluster is the default watch mode for pods. By default we watch the entire cluster.
+	DefaultWatchCluster = true
+)
+
+// Provider represents a k8s provider.
+type Provider struct {
+	logger logrus.FieldLogger
+
+	PodsInf             cache.SharedIndexInformer
+	annotationWhitelist []regexp.Regexp
+	labelWhitelist      []regexp.Regexp
```

Comment:
Do we need lists here? A regex can have branches `(A branch|B branch|something else)` - is that not good enough?

Reply:
It was separated because labels and annotations often have very different formats, and to logically allow two places for people to specify their regexes in the Viper config.

Annotation example: `gostatsd.atlassian.com/mytag=value`
Label example: `app=myapp`

Reply:
Sorry, let me rephrase that. Can we do this instead:
```go
	annotationFilter regexp.Regexp
	labelFilter      regexp.Regexp
```

Reply:
Yeah we could do that. I'll make it happen.